File: servercode/C++/rpi-universal.cpp
#include <iostream>
#include <string>
#include <cstring>
#include <unistd.h>
#include <arpa/inet.h>

const int PORT = 5000;
const std::string SERVER_PASSWORD = "11223344";

int main() {

    int server_fd = socket(AF_INET, SOCK_STREAM, 0);

    sockaddr_in serverAddr{};
    serverAddr.sin_family = AF_INET;
    serverAddr.sin_addr.s_addr = INADDR_ANY;
    serverAddr.sin_port = htons(PORT);

    bind(server_fd, (sockaddr*)&serverAddr, sizeof(serverAddr));

    listen(server_fd, 5);

    std::cout << "SERVER GESTART" << std::endl;

    while (true) {

        sockaddr_in clientAddr{};
        socklen_t clientSize = sizeof(clientAddr);

        int client = accept(server_fd,
                            (sockaddr*)&clientAddr,
                            &clientSize);

        std::cout << "Nieuwe client verbonden." << std::endl;

        char buffer[1024] = {0};

        int bytes = recv(client, buffer, sizeof(buffer), 0);

        if (bytes <= 0) {
            close(client);
            continue;
        }

        std::string password(buffer);

        while (!password.empty() &&
              (password.back() == '\n' || password.back() == '\r'))
            password.pop_back();

        if (password != SERVER_PASSWORD) {

            send(client,
                 "LOGIN_FAILED\n",
                 13,
                 0);

            close(client);
            continue;
        }

        send(client,
             "LOGIN_OK\n",
             9,
             0);

        while (true) {

            memset(buffer, 0, sizeof(buffer));

            bytes = recv(client,
                         buffer,
                         sizeof(buffer),
                         0);

            if (bytes <= 0)
                break;

            std::cout << buffer;

            std::string reply =
                "Raspberry Pi zegt: Hallo!\n";

            send(client,
                 reply.c_str(),
                 reply.size(),
                 0);
        }

        close(client);
    }

    close(server_fd);
}

File: servercode/Rust/rpi-universal.rs
use std::io::{BufRead, BufReader, Write};
use std::net::{TcpListener, TcpStream};

const PORT: u16 = 5000;
const SERVER_PASSWORD: &str = "11223344";

fn handle_client(mut stream: TcpStream) {
    println!("Nieuwe client verbonden.");

    let mut reader = BufReader::new(stream.try_clone().unwrap());

    let mut password = String::new();

    if reader.read_line(&mut password).is_err() {
        return;
    }

    password = password.trim().to_string();

    if password != SERVER_PASSWORD {
        let _ = stream.write_all(b"LOGIN_FAILED\n");
        return;
    }

    let _ = stream.write_all(b"LOGIN_OK\n");

    loop {
        let mut message = String::new();

        match reader.read_line(&mut message) {
            Ok(0) => break,
            Ok(_) => {
                println!("{}", message.trim());

                let _ = stream.write_all(
                    b"Raspberry PI server zegt: Hallo!\n"
                );
            }
            Err(_) => break,
        }
    }

    println!("Client verbroken.");
}

fn main() {
    let listener = TcpListener::bind(("0.0.0.0", PORT))
        .expect("Kan poort niet openen");

    println!("SERVER GESTART op poort {}", PORT);

    for stream in listener.incoming() {
        match stream {
            Ok(stream) => handle_client(stream),
            Err(e) => println!("Fout: {}", e),
        }
    }
}
